hs_composite_resource/models.py

```diff
@@ -52,24 +134,299 @@ def get_metadata_xml(self, pretty_print=True, include_format_elements=True):
+                # there may be single file aggregation in that folder for which xml documents
+                # need to be created
+                self._recreate_xml_docs_for_folder(aggregation_name, check_metadata_dirty=True)
+
+    def _recreate_xml_docs_for_folder(self, folder, check_metadata_dirty=False):
+        """Re-creates xml metadata and map documents associated with the specified folder.
+        If the *folder* represents an aggregation then map and metadata xml documents are
+        recreated only for that aggregation. Otherwise, xml documents are created for any
+        aggregation that may exist in the specified folder and its sub-folders.
+
+        :param  folder: folder for which xml documents need to be re-created
+        :param  check_metadata_dirty: if true, then xml files will be created only if the
+        aggregation metadata is dirty
+        """
+
+        # first check if the the folder represents an aggregation
+        try:
+            aggregation = self.get_aggregation_by_name(folder)
+            if check_metadata_dirty:
+                if aggregation.metadata.is_dirty:
+                    aggregation.create_aggregation_xml_documents()
+            else:
+                aggregation.create_aggregation_xml_documents()
+                # if we found an aggregation by the folder name that means this folder doesn't
+                # have any sub folders as multi-file aggregation folder can't have sub folders
+        except ObjectDoesNotExist:
+            # create xml map and metadata xml documents for all aggregations that exist
+            # in *folder* and its sub-folders
+            if not folder.startswith(self.file_path):
+                folder = os.path.join(self.file_path, folder)
+
+            res_file_objects = ResourceFile.list_folder(self, folder)
+            aggregations = []
+            for res_file in res_file_objects:
+                if res_file.has_logical_file and res_file.logical_file not in aggregations:
+                    aggregations.append(res_file.logical_file)
+
+            if check_metadata_dirty:
+                aggregations = [aggr for aggr in aggregations if aggr.metadata.is_dirty]
+            for aggregation in aggregations:
+                aggregation.create_aggregation_xml_documents()
+
+    def get_aggregation_by_name(self, name):
+            """Get an aggregation that matches the aggregation name specified by *name*
+            :param  name: name of the aggregation to find
+            :return an aggregation object if found
+            :raises ObjectDoesNotExist if no matching aggregation is found
+            """
+            for aggregation in self.logical_files:
+                if aggregation.aggregation_name == name:
+                    return aggregation
+
+            raise ObjectDoesNotExist("No matching aggregation was found for name:{}".format(name))
+
+    def recreate_aggregation_xml_docs(self, orig_aggr_name, new_aggr_name):
```

Comment:
Ok, I'm still wrapping my head around some of this xml metadata stuff but I feel like outlining all of these cases is over-complicating things.  If we think of the resource filesystem as a tree we can look at the node that is dirty and assume that everything below it is dirty as well as everything between it's branch and the root is dirty.  

Is this an incorrect model?  It doesn't seem to hold true to the use cases you're outlining here.

Reply:
When you say aggregations can nest, do you mean that aggregations can be placed inside of another aggregation?  If so, I believe my model still holds.

Reply:
I agree that this code is complicated. I wrote the cases outlined as part of the doc string as I was writing the code to handle all these cases. I did not see any harm in leaving those in doc string after I finished writing code for this function. I think my unit tests cover all these cases.

There may be a better way to write this logic. Most of the logic here is to find the what node is dirty. I guess I am not sure how the model you are proposing would work. If you can provide bit more details that may help.